server/request_handler.go

```diff
@@ -0,0 +1,127 @@
+	timeStarted     time.Time
+	defaultProxyUrl string
+	relaySigningKey *ecdsa.PrivateKey
+}
+
+func NewRpcRequestHandler(respw *http.ResponseWriter, req *http.Request, proxyUrl string, relaySigningKey *ecdsa.PrivateKey) *RpcRequestHandler {
+	return &RpcRequestHandler{
+		respw:           respw,
+		req:             req,
+		timeStarted:     Now(),
+		defaultProxyUrl: proxyUrl,
+		relaySigningKey: relaySigningKey,
+	}
+}
+
+func (r *RpcRequestHandler) process() {
+
+	// At end of request, log the time it needed
+	defer func() {
+		timeRequestNeeded := time.Since(r.timeStarted)
+		r.logger.log("request took %.6f sec", timeRequestNeeded.Seconds())
+	}()
+
+	ip := utils.GetIP(r.req)             // Fetch ip
+	origin := r.req.Header.Get("Origin") // Fetch origin
+	// Logger setup
+	r.logger = NewLogger(uuid.New().String(), ip, origin)
+	r.logger.log("POST request received")
+
+	// Validate if ip blacklisted
+	if IsBlacklisted(ip) {
+		r.logger.log("Blocked IP: %s", ip)
+		r._writeHeaderStatus(http.StatusUnauthorized)
+		return
+	}
+
+	// If users specify a proxy url in their rpc endpoint they can have their requests proxied to that endpoint instead of Infura
+	// e.g. https://rpc.flashbots.net?url=http://RPC-ENDPOINT.COM
+	customProxyUrl, ok := r.req.URL.Query()["url"]
+	if ok && len(customProxyUrl[0]) > 1 {
+		r.defaultProxyUrl = customProxyUrl[0]
+		r.logger.log("Using custom url: %s", r.defaultProxyUrl)
+	}
+
+	// Decode request JSON RPC
+	defer r.req.Body.Close()
+	body, err := ioutil.ReadAll(r.req.Body)
+	if err != nil {
+		r.logger.logError("failed to read request body: %v", err)
+		r._writeHeaderStatus(http.StatusBadRequest)
+		return
+	}
+
+	if len(body) == 0 {
+		r._writeHeaderStatus(http.StatusBadRequest)
+		return
+	}
+
+	// Parse JSON RPC payload
+	var jsonReq *types.JsonRpcRequest
+	if err = json.Unmarshal(body, &jsonReq); err != nil {
+		var jsonBatchReq []*types.JsonRpcRequest
+		if err = json.Unmarshal(body, &jsonBatchReq); err != nil {
+			r.logger.logError("Parse payload %v", err)
+			r._writeHeaderStatus(http.StatusBadRequest)
+			return
+		}
+		// Process batch request
+		r.processBatchRequest(jsonBatchReq)
+		return
+	}
+	// Process single request
+	r.processRequest(jsonReq)
+
+}
+
+// processRequest handles single request
+func (r *RpcRequestHandler) processRequest(jsonReq *types.JsonRpcRequest) {
+	// Handle single request
+	rpcReq := NewRpcRequest(r.logger, jsonReq, r.defaultProxyUrl, r.relaySigningKey)
+	res := rpcReq.ProcessRequest()
+	// Write response
+	r._writeRpcResponse(res)
+}
+
+// processBatchRequest handles multiple batch request
+func (r *RpcRequestHandler) processBatchRequest(jsonBatchReq []*types.JsonRpcRequest) {
+	batchCh := make(chan *types.JsonRpcResponse, len(jsonBatchReq)) // Chan to hold response from each go routine
+	for i := 0; i < cap(batchCh); i++ {
+		// Process each individual request
+		// Scatter worker
+		go func(rpcReq *types.JsonRpcRequest, ch chan *types.JsonRpcResponse) {
```

Comment:
The channel doesn't need to be passed into the function here.

```suggestion
		go func(rpcReq *types.JsonRpcRequest) {
```

Reply:
Done.